File: cpp/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerSimple.cpp
#include "CommandOptionsProcessor.h"
#include "ImageExporter.h"
#include "RaytracerExecutor.h"
#include "RaytracerParallelExecutor.h"
#include "RaytracerSerialExecutor.h"

#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/common/statistics/RaytraceStatistics.h"
#include "vsdk/toolkit/environment/camera/Camera.h"
#include "vsdk/toolkit/environment/camera/CameraSnapshot.h"
#include "vsdk/toolkit/environment/material/RendererConfiguration.h"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "vsdk/toolkit/environment/scene/SimpleSceneSnapshot.h"
#include "vsdk/toolkit/gui/feedback/ProgressMonitorConsole.h"
#include "vsdk/toolkit/io/geometry/ReaderMitScene.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/processing/StopWatch.h"
#include "vsdk/toolkit/render/raytracing/SimpleRaytracer.h"

#include <cstdlib>
#include <exception>
#include <cstdio>
#include "java/util/ArrayList.txx"

static const char* SCENE_SAMPLES_PATH = "../../../../etc/geometry/mitscenes/";
static const int ELAPSED_TIME_DECIMALS = 3;
static const int EXIT_CODE_READ_ERROR = -1;
static const int EXIT_CODE_IMAGE_ERROR = 1;
static const int EXIT_CODE_ARGUMENT_ERROR = 2;

static void optimizeRendererConfigurationForScene(SimpleScene* scene, RendererConfiguration* rendererConfiguration)
{
    bool hasTextures = false;
    bool hasNormalMaps = false;
    java::ArrayList<SimpleBody*>& bodies = scene->getSimpleBodies();

    for (long int i = 0; i < bodies.size(); i++) {
        SimpleBody* body = bodies[i];
        if ( body->getTexture() != 0 ) hasTextures = true;
        if ( body->getNormalMap() != 0 ) hasNormalMaps = true;
        if ( hasTextures && hasNormalMaps ) break;
    }

    rendererConfiguration->setTexture(hasTextures);
    rendererConfiguration->setBumpMap(hasNormalMaps);
}

static void offlineExecution(const java::String& fileName,
                             bool save,
                             const java::String& outputFileName,
                             bool parallel)
{
    SimpleScene scene;

    printf("Loading scene from %s: \n", fileName.c_str());

    try {
        ReaderMitScene readerMitScene;
        readerMitScene.importEnvironment(fileName.c_str(), &scene);
    }
    catch (const std::exception& e) {
        fprintf(stderr, "Error reading %s: %s\n", fileName.c_str(), e.what());
        fprintf(stderr, "There are scene samples on %s\n", SCENE_SAMPLES_PATH);
        std::exit(EXIT_CODE_READ_ERROR);
    }

    printf("Scene loaded OK!\n");

    RGBImageUncompressed resultingImage;
    Camera* activeCamera = scene.getActiveCamera();
    if ( !resultingImage.initNoFill((int)activeCamera->getViewportXSize(),
                                    (int)activeCamera->getViewportYSize()) ) {
        fprintf(stderr, "Error creating image!\n");
        std::exit(EXIT_CODE_IMAGE_ERROR);
    }

    ProgressMonitorConsole reporter;
    RendererConfiguration rendererConfiguration;
    optimizeRendererConfigurationForScene(&scene, &rendererConfiguration);

    SimpleRaytracer visualizationEngine;
    CameraSnapshot* cameraSnapshot =
        activeCamera->exportToCameraSnapshot(resultingImage.getXSize(), resultingImage.getYSize());
    SimpleSceneSnapshot* sceneSnapshot =
        scene.exportToSimpleSceneSnapshot(cameraSnapshot, scene.getActiveBackground());

    StopWatch clock;
    RaytracerExecutor* raytracerExecutor =
        parallel ? (RaytracerExecutor*)new RaytracerParallelExecutor() : (RaytracerExecutor*)new RaytracerSerialExecutor();

    clock.start();
    raytracerExecutor->run(&visualizationEngine,
                           &resultingImage,
                           &rendererConfiguration,
                           sceneSnapshot,
                           &reporter);
    clock.stop();

    printf("Image generated in %s seconds.\n",
        VSDK::formatDouble(clock.getElapsedRealTime(), ELAPSED_TIME_DECIMALS).c_str());
    RaytraceStatistics::printSummary();

    if ( save ) {
        ImageExporter imageExporter;
        if ( !imageExporter.exportImage(outputFileName, &resultingImage) ) {
            fprintf(stderr, "Error saving output image!\n");
            std::exit(EXIT_CODE_IMAGE_ERROR);
        }
    }

    delete raytracerExecutor;
    delete sceneSnapshot;
}

int main(int argc, char** argv)
{
    java::ArrayList<java::String> args;
    for (int i = 1; i < argc; i++) {
        args.add(argv[i]);
    }

    CommandOptionsProcessor options = CommandOptionsProcessor::process(args);
    if ( options.shouldShowHelp() ) {
        CommandOptionsProcessor::printUsage();
        if ( argc > 1 ) {
            return EXIT_CODE_ARGUMENT_ERROR;
        }
        return 0;
    }

    offlineExecution(options.getSceneFile(),
                     options.shouldSave(),
                     options.getOutputFile(),
                     options.shouldUseParallelExecutor());
    return 0;
}

File: typescript/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerSimple.ts
// Java classes
import { platformPrintln } from "@vitral/base";
import type { InputStream } from "@vitral/base";
import type { ArrayList } from "@vitral/base";
import { FileInputStream } from "@vitral/fs";

// Vitral classes
import { RendererConfiguration } from "@vitral/base";
import { StopWatch } from "@vitral/base";
import { VSDK } from "@vitral/base";
import { RaytraceStatistics } from "@vitral/base";
import { RGBImageUncompressed } from "@vitral/base";
import type { Camera } from "@vitral/base";
import type { CameraSnapshot } from "@vitral/base";
import type { SimpleBody } from "@vitral/base";
import { SimpleScene } from "@vitral/base";
import type { SimpleSceneSnapshot } from "@vitral/base";
import { ProgressMonitorConsole } from "@vitral/base";
import { SimpleRaytracer } from "@vitral/base";
import { ReaderMitScene } from "@vitral/fs";

import { CommandOptionsProcessor } from "./CommandOptionsProcessor.js";
import { ImageExporter } from "./ImageExporter.js";
import type { RaytracerExecutor } from "./RaytracerExecutor.js";
import { RaytracerParallelExecutor } from "./RaytracerParallelExecutor.js";
import { RaytracerSerialExecutor } from "./RaytracerSerialExecutor.js";

export class RaytracerSimple {
    private static readonly SCENE_SAMPLES_PATH = "../../../../etc/geometry/mitscenes/";
    private static readonly ELAPSED_TIME_DECIMALS = 3;
    private static readonly EXIT_CODE_READ_ERROR = -1;
    private static readonly EXIT_CODE_IMAGE_ERROR = 1;
    private static readonly EXIT_CODE_ARGUMENT_ERROR = 2;

    private readonly scene: SimpleScene;

    public constructor() {
        this.scene = new SimpleScene();
    }

    private optimizeRendererConfigurationForScene(rendererConfiguration: RendererConfiguration): void {
        let hasTextures = false;
        let hasNormalMaps = false;
        const bodies: ArrayList<SimpleBody> = this.scene.getSimpleBodies();

        for (let i = 0; i < bodies.size(); i++) {
            const body: SimpleBody = bodies.get(i);
            if (body.getTexture() !== null) {
                hasTextures = true;
            }
            if (body.getNormalMap() !== null) {
                hasNormalMaps = true;
            }
            if (hasTextures && hasNormalMaps) {
                break;
            }
        }

        rendererConfiguration.setTexture(hasTextures);
        rendererConfiguration.setBumpMap(hasNormalMaps);
    }

    private async offlineExecution(
        fileName: string,
        save: boolean,
        outputFileName: string,
        parallel: boolean,
    ): Promise<void> {
        let resultingImage: RGBImageUncompressed;

        //- 1. Import the scene from scene description file to RAM -----
        console.log("Loading scene from " + fileName + ": ");
        let is: InputStream | null = null;
        try {
            is = new FileInputStream(fileName);
            const readerMitScene: ReaderMitScene = new ReaderMitScene();
            readerMitScene.importEnvironment(is, this.scene);
        } catch {
            console.error("Error reading " + fileName);
            console.error("There are scene samples on " + RaytracerSimple.SCENE_SAMPLES_PATH);
            process.exit(RaytracerSimple.EXIT_CODE_READ_ERROR);
        } finally {
            // `try (InputStream is = ...)`: the reader already closes the
            // stream on the success path, and this closes it on the error one.
            is?.close();
        }
        console.log("Scene loaded OK!");

        //- 2. Create an empty image --------------------------------------
        resultingImage = new RGBImageUncompressed();
        const activeCamera: Camera = this.scene.getActiveCamera();
        if (
            !resultingImage.initNoFill(
                Math.trunc(activeCamera.getViewportXSize()),
                Math.trunc(activeCamera.getViewportYSize()),
            )
        ) {
            console.error("Error creating image!");
            process.exit(RaytracerSimple.EXIT_CODE_IMAGE_ERROR);
        }

        //- 3. Process the image from the scene data structure -----------
        const reporter: ProgressMonitorConsole = new ProgressMonitorConsole();
        const rendererConfiguration: RendererConfiguration = new RendererConfiguration();
        this.optimizeRendererConfigurationForScene(rendererConfiguration);

        const visualizationEngine: SimpleRaytracer = new SimpleRaytracer();
        const cameraSnapshot: CameraSnapshot = activeCamera.exportToCameraSnapshot(
            resultingImage.getXSize(),
            resultingImage.getYSize(),
        );
        const sceneSnapshot: SimpleSceneSnapshot = this.scene.exportToSimpleSceneSnapshot(
            cameraSnapshot,
            this.scene.getActiveBackground(),
        );

        const clock: StopWatch = new StopWatch();
        const raytracerExecutor: RaytracerExecutor = parallel
            ? new RaytracerParallelExecutor(fileName, new URL("./RaytracerTileWorker.js", import.meta.url))
            : new RaytracerSerialExecutor();

        clock.start();
        await raytracerExecutor.run(
            visualizationEngine,
            resultingImage,
            rendererConfiguration,
            sceneSnapshot,
            reporter,
        );
        clock.stop();

        platformPrintln(
            "Image generated in " +
                VSDK.formatDouble(clock.getElapsedRealTime(), RaytracerSimple.ELAPSED_TIME_DECIMALS) +
                " seconds.",
        );
        RaytraceStatistics.printSummary();

        //- 4. Export resulting image to an image file --------------------
        if (save) {
            const imageExporter: ImageExporter = new ImageExporter();
            if (!imageExporter.export(outputFileName, resultingImage)) {
                console.error("Error saving output image!");
                process.exit(RaytracerSimple.EXIT_CODE_IMAGE_ERROR);
            }
        }
    }

    public static async main(args: string[]): Promise<void> {
        const instance: RaytracerSimple = new RaytracerSimple();
        const options: CommandOptionsProcessor = CommandOptionsProcessor.process(args);
        if (options.shouldShowHelp()) {
            CommandOptionsProcessor.printUsage();
            if (args.length > 0) {
                process.exit(RaytracerSimple.EXIT_CODE_ARGUMENT_ERROR);
            }
        }
        await instance.offlineExecution(
            options.getSceneFile(),
            options.shouldSave(),
            options.getOutputFile(),
            options.shouldUseParallelExecutor(),
        );
    }
}

await RaytracerSimple.main(process.argv.slice(2));
